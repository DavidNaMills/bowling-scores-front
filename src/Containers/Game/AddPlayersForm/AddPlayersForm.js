import React, { useState, useEffect } from 'react';
// import { useSelector } from 'react-redux';
import Button from '../../../Components/StandAloneComponents/Button/Button';
import Table from '../../../Components/Table/Table';
import Input from '../../../Components/Form/elements/Input/Input';
import Title from '../../../Components/StandAloneComponents/Title/Title';

import { playerTableStyle } from '../playerTableStyle';
import tableParser from '../../../helpers/tableGamesDataModifier/tableGamesDataModifier';

import useAddPlayer from '../../../Hooks/useAddPlayer/useAddPlayer';
import useColorSelection from '../../../Hooks/useColorSelection/useColorSelection';

const tHeaders = ['Player', 'Average', 'Pinfall'];



const AddPlayersForm = ({ title, playerSelect, onClose, isNew = false, close = () => { } }) => {
    // const liveGame = useSelector(state => state.liveGame);
    const {
        liveGame,   //temp structure
        newPlayer,
        setName,    //only sets player name
        addColor,    //sets color
        addPlayer,
        commitGame,
    } = useAddPlayer(isNew);

    const { color, randomColor, showColorPickerComponent } = useColorSelection();
    const [showSelect, setShowSelect] = useState(false);


    useEffect(() => {
        if (color) {
            addColor(color);
        }
    }, [color]);


    return (
        <div>
            <Title label={title} ttlType='sub' />
            <Input
                value={newPlayer.name}
                id='name'
                label='Players Name'
                changed={setName}
            />

            {
                showSelect && showColorPickerComponent()

            }
            <Button label={showSelect ? 'Close' : 'Select Color'} click={() => setShowSelect(prev => !prev)} />
            <Button label='Random Colour' click={randomColor} />
            {newPlayer.color &&
                <div
                    style={{
                        marginTop: '10px',
                        marginBottom: '10px',
                        height: '30px',
                        textAlign: 'center',
                        ...playerTableStyle(newPlayer.color)
                    }}
                >{newPlayer.name}</div>}


            <Button label='Add Player' click={addPlayer} />
            {isNew && <Button label='Start Game' click={() => commitGame(close)} />}
            <Table
                data={{
                    headers: tHeaders,
                    rows: tableParser(liveGame)
                }}
                selectRow={playerSelect}
                showRowNum
                caption='Click player for more details'
            />
            <Button label='Close' click={onClose} />
        </div>
    )
}

export default AddPlayersForm;