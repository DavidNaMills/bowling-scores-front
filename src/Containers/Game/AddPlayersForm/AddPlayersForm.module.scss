.addPlayers{
    &__halfBtn{
        width: 50%;
    }

    &__fullBtn{
        width: 100%;
    }

    &__inline{
        display: flex;
        justify-content: space-between;
    }
}