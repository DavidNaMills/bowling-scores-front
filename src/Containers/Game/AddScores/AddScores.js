import React, { useEffect } from 'react';
import CombinedInput from './CombinedInput/CombinedInput';
import objectToArray from '../../../helpers/objectToArray/objectToArray'
import Button from '../../../Components/StandAloneComponents/Button/Button';
import Title from '../../../Components/StandAloneComponents/Title/Title';

import useScoreInput from '../../../Hooks/useScoreInput/useScoreInput';
import useDispatchHook from '../../../Hooks/useDispatchHook/useDispatchHook';

const AddScores = ({ players, close }) => {
    const { values, addValue } = useScoreInput(players);
    const {addNewGameDispatch} = useDispatchHook();

    const updateScore = (e) => {
        e.preventDefault();
        let touched = false;

        for (let k in values) {
            if (values[k].score > 0) touched = true;
        }
        if (touched) {
            addNewGameDispatch(values);
            console.log(values);
            close();
        } else {
            alert('must add at least 1 score')
        }
    }

    useEffect(() => { }, [values]);

    return (
        <div>
            <Title label='Add Scores' ttlType='sub' />
            {
                values ?
                    <React.Fragment>
                        <form onSubmit={updateScore}>
                            {
                                objectToArray(players).map((x, i) =>
                                    <CombinedInput
                                        key={i}
                                        name={x.config.name}
                                        color={x.config.color}
                                        updateScore={addValue}
                                        id={x.id}
                                        value={values[x.id].score}
                                    />
                                )}
                            <Button click={() => { }} isFull label='Save Scores' />
                        </form>

                        <div>
                            <Button isFull click={close} label='Close' />
                        </div>
                    </React.Fragment>
                    : <p>Please wait</p>
            }
        </div>
    )
}

export default AddScores;