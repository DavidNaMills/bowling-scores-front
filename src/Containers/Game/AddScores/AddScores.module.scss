@import '../../../styles/colors.scss';

.addScores{
    &__form{
        width: 75vw;
        max-width: 400px;
        border: solid $popup_color 2px;
    }
}