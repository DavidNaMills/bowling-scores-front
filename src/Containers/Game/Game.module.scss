.test {
    background: purple;
    color: yellow;
}