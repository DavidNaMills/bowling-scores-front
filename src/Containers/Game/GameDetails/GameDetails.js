import React from 'react';
import Button from '../../../Components/StandAloneComponents/Button/Button';
import Table from '../../../Components/Table/Table';
import GameChart from '../../../Components/GameChart/GameChart';

import tableParser from '../../../helpers/tableGamesDataModifier/tableGamesDataModifier';
import chartParser from '../../../helpers/chartDataModifier/chartDataModifier';

const tHeaders = ['Player', 'Average', 'Pinfall'];

const GameDetails = ({ liveGame, addNewPlayers, playerSelect, addScores, newGame }) => {
    return (
        <div>
            <Button click={newGame} label='Start new game' />
            <Button click={addScores} label='Add Scores' />

            <GameChart
                players={liveGame.players}
                data={chartParser(liveGame)}
            />
            <Table
                data={{
                    headers: tHeaders,
                    rows: tableParser(liveGame)
                }}
                showRowNum
                selectRow={playerSelect}
                caption='Click player for more details'
            />
            <Button click={addNewPlayers} label='Add Players' />
        </div>
    )
}


export default GameDetails;