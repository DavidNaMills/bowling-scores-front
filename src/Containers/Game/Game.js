import React, { useState, useEffect } from 'react';

import { useSelector, useDispatch } from 'react-redux';
import * as actions from '../../store/allActions';

import GameDetails from './GameDetails/GameDetails';
import AddScores from './AddScores/AddScores';
import AddPlayersForm from './AddPlayersForm/AddPlayersForm';

import Title from '../../Components/StandAloneComponents/Title/Title';

import classes from './Game.module.scss';


const showDefault = {
    gameDetails: true,
    addPlayers: false,
    addscores: false
}

const Game = (props) => {
    console.log('[Game]');

    const dispatch = useDispatch();
    const liveGame = useSelector(state => state.liveGame);

    const createNewGame = () => dispatch(actions.initGame());
    const addNewPlayerDispatch = (data) => dispatch(actions.addPlayer(data));
    const addNewGameDispatch = (data) => dispatch(actions.addNewGame(data));

    const [showWhich, setWhich] = useState(showDefault);

    // useEffect(() => { }, [liveGame])

    const playerSelect = (id) => {
        setWhich(showDefault);
        props.history.push({
            pathname: '/player',
            state: {
                id
            }
        })
    }

    const changeFromGame = (id) => {
        const base = {
            gameDetails: false,
            addPlayers: false,
            addscores: false
        };
        setWhich({
            ...base,
            [id]: true
        })
    }

    const addNewGame = (data) => {
        addNewGameDispatch(data)
    }

    const buildGameDetails = () => (
        <GameDetails
            liveGame={liveGame}
            addNewPlayers={() => changeFromGame('addPlayers')}
            playerSelect={playerSelect}
            addScores={() => changeFromGame('addscores')}
        />
    )

    const buildAddPlayersForm = () => (
        <AddPlayersForm
            addNewPlayer={addNewPlayerDispatch}
            liveGame={liveGame}
            playerSelect={playerSelect}
            onClose={() => setWhich(showDefault)}
        />
    )

    const buildAddScores = () => (
        <AddScores
            players={liveGame.players}
            addGame={addNewGame}
            close={() => setWhich(showDefault)}
        />
    )

    return (
        <div>
            <Title ttlType='section' label='Play' />
            {showWhich.gameDetails && buildGameDetails()}
            {showWhich.addPlayers && buildAddPlayersForm()}
            {showWhich.addscores && buildAddScores()}

        </div>
    )
}
{/* <GameDetails
    addNewPlayers={addNewPlayers}
    liveGame={liveGame}
    playerSelect={playerSelect}
    onClose={onClose}
/> */}

export default Game;


/**
 * <Button click={startNewGame} label='New Game'/>

            <GameChart
                players={liveGame.players}
                data={chartParser(liveGame)}
            />
            <Table
                data={{
                    headers: tHeaders,
                    rows: tableParser(liveGame)
                }}
                showRowNum
                selectRow={onClick}
                caption='Click player for more details'
            />
 */