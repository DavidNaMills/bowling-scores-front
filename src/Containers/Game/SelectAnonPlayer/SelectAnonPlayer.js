import React, { useState } from 'react';
import classes from './SelectAnonPlayer.module.scss';
import spacing from '../../../styles/shared/spacing.module.scss';


import Title from '../../../Components/StandAloneComponents/Title/Title';
import Table from '../../../Components/Table/Table';
import Button from '../../../Components/StandAloneComponents/Button/Button';

import pinfallSort from '../../../helpers/pinfallSort/pinfallSort';
import tableParser from '../../../helpers/tableGamesDataModifier/tableGamesDataModifier';
import { playerTableStyle } from '../../../Containers/Game/playerTableStyle';
import useLocalStorage from '../../../Hooks/useLocalStorage/useLocalStorage';

const label1 = 'Please select the name you played as'

const tHeaders = ['Player', 'Average', 'Pinfall'];

const SelectAnonPlayer = (props) => {
    const { removeFromStorage } = useLocalStorage();
    const {
        game,
        cancelClick,
        submitClick
    } = props;

    const [sel, setSel] = useState(null)

    return (
        <div>
            <Title label={label1} />
            <Table
                data={{
                    headers: tHeaders,
                    rows: pinfallSort(tableParser(game))
                }}
                caption='Select Yourself'
                selectRow={setSel}
            />

            {sel &&
                    <p className={spacing.largeExtra} style={{ ...playerTableStyle(game.players[sel].color), padding: '10px', textAlign: 'center' }}>
                        {`Selected Player: ${game.players[sel].name}`}
                    </p>
            }

            <div className={spacing.largeExtra}>
                <Button click={() => submitClick(sel)} label={'Submit'} isDisabled={!sel ? true : false} type='lightgreen' isFull />
            </div>
            <div className={spacing.extra}>
                <Button click={()=>{
                    removeFromStorage();
                    cancelClick();
                }
                } label={'Cancel'} type='lightred' isFull />
            </div>
        </div>
    )
}

export default SelectAnonPlayer;