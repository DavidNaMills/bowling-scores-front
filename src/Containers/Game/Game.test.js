import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import {act} from 'react-dom/test-utils';

import Game from './Game';
import GameDetails from './GameDetails/GameDetails';
import AddScores from './AddScores/AddScores';
import AddPlayersForm from './AddPlayersForm/AddPlayersForm';
import SelectAnonPlayer from './SelectAnonPlayer/SelectAnonPlayer';

jest.mock('../../Hooks/useShowHook/useShow');
import useShowHook from '../../Hooks/useShowHook/useShow';

jest.mock('react-redux');
import { useSelector } from 'react-redux';

import Spinner from '../../Components/Spinner/Spinner';

const mockState = {
    liveGame: {
        isLoading: false,
        games: {
            1: { 'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", score: 12 } },
            2: { 'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", score: 24 } },
            3: { 'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", score: 154 } },
        },
        players: {
            'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", color: "17, 127, 14", id: "debb53c3-ad2a-4093-b89b-2a5025416f6c" }
        }
    }
};

afterEach(() => {
    jest.clearAllMocks();
});

describe('<Game/> component test suite', () => {
    describe('creation tests', () => {

        beforeEach(() => {
            useShowHook.mockImplementation(() => ({
                isShow: {
                    gameDetails: true,
                    addPlayers: false,
                    addscores: false,
                    newGame: false,
                    playerSelect: false
                },
                changeShow: jest.fn()
            }));

            useSelector.mockImplementation(() => ({
                user: { token: 'fdsfdsa', user: { name: 'david', color: '123,456,789' } },
                liveGame: { isLoading: true }
            }));
        });

        useShowHook.mockImplementation(() => ({
            isShow: {
                gameDetails: true,
                addPlayers: false,
                addscores: false,
                newGame: false,
                playerSelect: false
            },
            changeShow: jest.fn()
        }))

        it('matches snapshot', () => {
            const comp = renderer.create(<Game />);
            const app = comp.toJSON();
            expect(app).toMatchSnapshot();
        });

        it('exists', () => {
            const wrapper = shallow(<Game />);
            expect(wrapper.exists).toBeTruthy();
        });
    });

    describe('render views tests', () => {
        const mockState = {
            user: { token: 'fdsfdsa', user: { name: 'david', color: '123,456,789' } },
            liveGame: { isLoading: false }
        }

        it('displays a spinner if isLoading is true', () => {
            useShowHook.mockImplementation(() => ({
                isShow: {
                    gameDetails: true,
                    addPlayers: false,
                    addscores: false,
                    newGame: false,
                    playerSelect: false
                },
                changeShow: jest.fn()
            }));

            useSelector.mockImplementation(() => ({
                user: { token: 'fdsfdsa', user: { name: 'david', color: '123,456,789' } },
                liveGame: { isLoading: true }
            }));
            const wrapper = mount(<Game />);

            expect(wrapper.find(Spinner).length).toBe(1);
        });

        // gameDetails displays GameDetails
        it('renders <GameDetails/> component', () => {
            useShowHook.mockImplementation(() => ({
                isShow: {
                    gameDetails: true,
                    addPlayers: false,
                    addscores: false,
                    newGame: false,
                    playerSelect: false
                },
                changeShow: jest.fn()
            }));

            useSelector.mockImplementation(() => (mockState));
            const wrapper = mount(<Game />);

            expect(wrapper.find(GameDetails).length).toBe(1);
        });

        it('renders <AddPlayersForm/>', ()=>{
            useShowHook.mockImplementation(() => ({
                isShow: {
                    gameDetails: false,
                    addPlayers: true,
                    addscores: false,
                    newGame: false,
                    playerSelect: false
                },
                changeShow: jest.fn()
            }));

            useSelector.mockImplementation(() => (mockState));
            const wrapper = mount(<Game />);

            expect(wrapper.find(AddPlayersForm).length).toBe(1);
        });

        it('renders <AddScores/>', ()=>{
            useShowHook.mockImplementation(() => ({
                isShow: {
                    gameDetails: false,
                    addPlayers: false,
                    addscores: true,
                    newGame: false,
                    playerSelect: false
                },
                changeShow: jest.fn()
            }));

            useSelector.mockImplementation(() => (mockState));
            const wrapper = mount(<Game />);

            expect(wrapper.find(AddScores).length).toBe(1);
        });
        

        it.skip('renders <SelectAnonPlayer/>', ()=>{
            // FIXME: 

            useShowHook.mockImplementation(() => ({
                isShow: {
                    gameDetails: false,
                    addPlayers: false,
                    addscores: false,
                    newGame: false,
                    playerSelect: true
                },
                changeShow: jest.fn()
            }));

            useSelector.mockImplementation(() => (mockState));

            const wrapper = mount(<Game />);
            act(()=>{
                wrapper.gameRef.current= {
                    games: {
                        "1": { 'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", score: "12" } },
                        "2": { 'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", score: "24" } },
                        "3": { 'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", score: "154" } },
                    },
                    players: {
                        'debb53c3-ad2a-4093-b89b-2a5025416f6c': { name: "gfdgfds", color: "17, 127, 14", id: "debb53c3-ad2a-4093-b89b-2a5025416f6c" }
                    }
                };
            });
            expect(wrapper.find(SelectAnonPlayer).length).toBe(1);

        });
        

        // playerSelect displays SelectAnonPlayer
    });

});

//TODO: test the logic: move into own hook or seperate function?
//

// game exists in localstorage
    // parsed.logged && ! show && user token && show && parsed.Id === user._id  === popup1
    // parsed.logged && ! show && user token && show && parsed.Id !== user._id  === cancel
    // parsed.logged && ! show && !user token === popup2

    // !parsed.logged && user token === popup3
    // !parsed.logged && ! user token === popup1
// NewGame
