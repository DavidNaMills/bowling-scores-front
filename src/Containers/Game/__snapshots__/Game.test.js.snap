// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> component test suite creation tests matches snapshot 1`] = `
<div>
  <div
    className="css-mnv97b"
  >
    <div
      className="css-42nwui"
    />
    <div
      className="css-5z299x"
    />
    <div
      className="css-4vijxx"
    />
    <div
      className="css-1ndb4r1"
    />
    <div
      className="css-1mvde9j"
    />
    <div
      className="css-fyec2"
    />
  </div>
  <div
    className="title"
  >
    <h2
      className="title__section"
    >
      Bowling Scores: Play
    </h2>
  </div>
  <div
    className="contentContainer"
  >
    <div
      className="btns"
    >
      <button
        className="button button__isFull button__blue"
        disabled={false}
        onClick={[Function]}
      >
        Add Scores
      </button>
    </div>
    <div
      className="title"
    >
      <h3
        className="title__sub"
      >
        Invalid-Date
      </h3>
    </div>
    <div
      className="gameDetails__center"
    >
      <div>
        <div
          className="recharts-wrapper"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "default",
              "height": 350,
              "position": "relative",
              "width": 800,
            }
          }
        >
          <svg
            className="recharts-surface"
            height={350}
            version="1.1"
            viewBox="0 0 800 350"
            width={800}
          >
            <defs>
              <clipPath
                id="recharts1-clip"
              >
                <rect
                  height={315}
                  width={755}
                  x={40}
                  y={5}
                />
              </clipPath>
            </defs>
            <g
              className="recharts-cartesian-grid"
            >
              <g
                className="recharts-cartesian-grid-horizontal"
              >
                <line
                  fill="none"
                  height={315}
                  stroke="#ccc"
                  strokeDasharray="3 3"
                  width={755}
                  x={40}
                  x1={40}
                  x2={795}
                  y={5}
                  y1={5}
                  y2={5}
                />
                <line
                  fill="none"
                  height={315}
                  stroke="#ccc"
                  strokeDasharray="3 3"
                  width={755}
                  x={40}
                  x1={40}
                  x2={795}
                  y={5}
                  y1={320}
                  y2={320}
                />
              </g>
              <g
                className="recharts-cartesian-grid-vertical"
              >
                <line
                  fill="none"
                  height={315}
                  stroke="#ccc"
                  strokeDasharray="3 3"
                  width={755}
                  x={40}
                  x1={40}
                  x2={40}
                  y={5}
                  y1={5}
                  y2={320}
                />
                <line
                  fill="none"
                  height={315}
                  stroke="#ccc"
                  strokeDasharray="3 3"
                  width={755}
                  x={40}
                  x1={795}
                  x2={795}
                  y={5}
                  y1={5}
                  y2={320}
                />
              </g>
            </g>
            <g
              className="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
            >
              <line
                className="recharts-cartesian-axis-line"
                fill="none"
                height={315}
                stroke="#666"
                width={40}
                x={0}
                x1={40}
                x2={40}
                y={5}
                y1={5}
                y2={320}
              />
              <g
                className="recharts-cartesian-axis-ticks"
              />
            </g>
          </svg>
          <div
            className="recharts-tooltip-wrapper"
            style={
              Object {
                "MozTransform": "translate(undefinedpx, undefinedpx)",
                "OTransform": "translate(undefinedpx, undefinedpx)",
                "WebkitTransform": "translate(undefinedpx, undefinedpx)",
                "msTransform": "translate(undefinedpx, undefinedpx)",
                "pointerEvents": "none",
                "position": "absolute",
                "top": 0,
                "transform": "translate(undefinedpx, undefinedpx)",
                "visibility": "hidden",
              }
            }
          >
            <div
              className="recharts-default-tooltip"
              style={
                Object {
                  "backgroundColor": "#fff",
                  "border": "1px solid #ccc",
                  "margin": 0,
                  "padding": 10,
                  "whiteSpace": "nowrap",
                }
              }
            >
              <p
                className="recharts-tooltip-label"
                style={
                  Object {
                    "margin": 0,
                  }
                }
              >
                
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="btns"
    >
      <div>
        <p
          className="caption"
        >
          Click player for more details
        </p>
        <table
          className="table"
        >
          <thead>
            <tr>
              <td />
              <td>
                Player
              </td>
              <td>
                Average
              </td>
              <td>
                Pinfall
              </td>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
    <div
      className="largeExtra"
    >
      <button
        className="button button__isFull button__blue"
        disabled={false}
        onClick={[Function]}
      >
        Add Players
      </button>
      <div
        className="extra"
      >
        <button
          className="button button__darkGreen"
          disabled={false}
          onClick={[Function]}
        >
          Start new game
        </button>
      </div>
    </div>
  </div>
</div>
`;
